### PR TITLE
Add ConvolutionalNetwork and LossConfig classes

### PR DESCRIPTION
**Pull request by mcgibbon**
We are working to add a convolutional model to fv3fit. This PR adds some building blocks to make that happen, particularly new dataclasses ConvolutionalNetwork and LossConfig which can be configured from yaml.

LossConfig duplicates functionality present in slightly different form in two places - the Dense loss, which applies per-feature scaling, and the PrecipitativeModel loss, which scales based on the standard deviation across all features after removing the feature mean. It also adds tests for these different loss options. This new configuration class is assumed to be the loss for a single variable, so the interface for weighting is greatly simplified. Ideally, we should refactor the existing code to use this new configuration class in a later PR.

Added public API:
- Added ConvolutionalNetwork class to configure a sequence of convolutional layers
- Added LossConfig class to configure a loss function for a single variable
- Added numpy_print_precision context manager to `fv3fit.testing`

Significant internal changes:
- Updated unit test code to include "x" and "y" dimensions in test data

- [x] Tests added

Resolves #<github issues> [JIRA-TAG]

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/fv3fit/fv3fit/keras/_models/shared/loss.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            tf.math.reduce_mean(tf.math.square((y_pred - y_true) / std), axis=0)
+        )
+
+    return custom_loss
+
+
+def _standard_scaled_mae(std):
+    std = tf.constant(std, dtype=std.dtype)
+
+    def custom_loss(y_true, y_pred):
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/shared/loss.py:20]: Nit- standard and uniform_standard functions handle the custom loss differently- can we just handle them both the same way for consistency?

File: external/fv3fit/fv3fit/keras/_models/shared/loss.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    """
+    Attributes:
+        loss_type: one of "mse" or "mae"
+        scaling: "standard" corresponds to scaling each feature's lossby
+            its scale, "standard_uniform" corresponds to scaling
+            each feature's loss by the mean of all feature scales, where
+            the scale is variance for MSE loss or standard deviation
+            for MAE loss
+        weight: A scaling factor by which to modify this loss
+    """
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/shared/loss.py:56]: The `DenseModel` has a `normalize_loss` option that defaults to True- I don't know if we've ever used the `False` option, but you may want to keep it in the `LossConfig` and add a test.

File: external/fv3fit/fv3fit/keras/_models/shared/convolutional_network.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            config: configuration for convolutional network
+            label: inserted into layer names, if this function is used multiple times
+                to build one network you must provide a different label each time
+        
+        Returns:
+            tensors resulting from the requested convolutional network, each
+            tensor has the same dimensionality as the input tensor but will have fewer
+            points along the x- and y-dimensions, due to convolutions
+        """
+        if self.depth < 1:
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/convolutional_network.py:81]: Yes, I'll refactor to use that pattern.

File: external/fv3fit/fv3fit/keras/_models/shared/loss.py (hunk 1 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/loss.py:20]: What do you mean when you say they "handle the custom loss differently"? The configuration is meant to change the behavior of the loss function so that it uses individual std values or the same std value for all levels/features.

File: external/fv3fit/fv3fit/keras/_models/shared/loss.py (hunk 2 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/loss.py:56]: I decided not to implement "no normalization" because we've never had a use case for it or even discussed possibly using it. When I added it to DenseModel I probably shouldn't have because of YAGNI.

File: external/fv3fit/fv3fit/keras/_models/shared/loss.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                loss = _uniform_scaled_mse(std)
+            elif self.scaling == "standard":
+                loss = _standard_scaled_mse(std)
+        elif self.loss_type == "mae":
+            if self.scaling == "standard_uniform":
+                loss = _uniform_scaled_mae(std)
+            elif self.scaling == "standard":
+                loss = _standard_scaled_mae(std)
+        else:
+            raise NotImplementedError(f"loss_type {self.loss_type} is not implemented")
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/loss.py:83]: I should refactor this as well.

File: external/fv3fit/tests/keras/test_convolutional_network.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        pytest.param(
+            (3, 11, 15, 2),
+            3,
+            3,
+            5,
+            (3, 7, 11),
+            id="two_convolutions_different_input_shape",
+        ),
+        pytest.param(
+            (3, 10, 10, 2), 3, 2, 10, (3, 8, 8), id="one_convolution_more_kernel_size"
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/tests/keras/test_convolutional_network.py:47]: Will fix, "larger filter" above should say "larger kernel" and this one should say "more filters"
